// ngx-image-cropper/bundles/ngx-image-cropper.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/platform-browser'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-image-cropper', ['exports', '@angular/core', '@angular/platform-browser', '@angular/common'], factory) :
    (factory((global['ngx-image-cropper'] = {}),global.ng.core,global.ng.platformBrowser,global.ng.common));
}(this, (function (exports,core,platformBrowser,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var ImageUtils = (function () {
        function ImageUtils() {
        }
        /**
         * @param {?} imageBase64
         * @return {?}
         */
        ImageUtils.getOrientation = /**
         * @param {?} imageBase64
         * @return {?}
         */
            function (imageBase64) {
                /** @type {?} */
                var view = new DataView(this.base64ToArrayBuffer(imageBase64));
                if (view.getUint16(0, false) != 0xFFD8) {
                    return -2;
                }
                /** @type {?} */
                var length = view.byteLength;
                /** @type {?} */
                var offset = 2;
                while (offset < length) {
                    if (view.getUint16(offset + 2, false) <= 8)
                        return -1;
                    /** @type {?} */
                    var marker = view.getUint16(offset, false);
                    offset += 2;
                    if (marker == 0xFFE1) {
                        if (view.getUint32(offset += 2, false) != 0x45786966) {
                            return -1;
                        }
                        /** @type {?} */
                        var little = view.getUint16(offset += 6, false) == 0x4949;
                        offset += view.getUint32(offset + 4, little);
                        /** @type {?} */
                        var tags = view.getUint16(offset, little);
                        offset += 2;
                        for (var i = 0; i < tags; i++) {
                            if (view.getUint16(offset + (i * 12), little) == 0x0112) {
                                return view.getUint16(offset + (i * 12) + 8, little);
                            }
                        }
                    }
                    else if ((marker & 0xFF00) != 0xFF00) {
                        break;
                    }
                    else {
                        offset += view.getUint16(offset, false);
                    }
                }
                return -1;
            };
        /**
         * @param {?} imageBase64
         * @return {?}
         */
        ImageUtils.base64ToArrayBuffer = /**
         * @param {?} imageBase64
         * @return {?}
         */
            function (imageBase64) {
                imageBase64 = imageBase64.replace(/^data\:([^\;]+)\;base64,/gmi, '');
                /** @type {?} */
                var binaryString = atob(imageBase64);
                /** @type {?} */
                var len = binaryString.length;
                /** @type {?} */
                var bytes = new Uint8Array(len);
                for (var i = 0; i < len; i++) {
                    bytes[i] = binaryString.charCodeAt(i);
                }
                return bytes.buffer;
            };
        /**
         * @param {?} srcBase64
         * @param {?} srcOrientation
         * @param {?} callback
         * @return {?}
         */
        ImageUtils.resetOrientation = /**
         * @param {?} srcBase64
         * @param {?} srcOrientation
         * @param {?} callback
         * @return {?}
         */
            function (srcBase64, srcOrientation, callback) {
                /** @type {?} */
                var img = new Image();
                img.onload = function () {
                    /** @type {?} */
                    var width = img.width;
                    /** @type {?} */
                    var height = img.height;
                    /** @type {?} */
                    var canvas = document.createElement('canvas');
                    /** @type {?} */
                    var ctx = canvas.getContext('2d');
                    if (ctx) {
                        if (4 < srcOrientation && srcOrientation < 9) {
                            canvas.width = height;
                            canvas.height = width;
                        }
                        else {
                            canvas.width = width;
                            canvas.height = height;
                        }
                        ImageUtils.transformCanvas(ctx, srcOrientation, width, height);
                        ctx.drawImage(img, 0, 0);
                        callback(canvas.toDataURL());
                    }
                    else {
                        callback(srcBase64);
                    }
                };
                img.src = srcBase64;
            };
        /**
         * @param {?} ctx
         * @param {?} orientation
         * @param {?} width
         * @param {?} height
         * @return {?}
         */
        ImageUtils.transformCanvas = /**
         * @param {?} ctx
         * @param {?} orientation
         * @param {?} width
         * @param {?} height
         * @return {?}
         */
            function (ctx, orientation, width, height) {
                switch (orientation) {
                    case 2:
                        ctx.transform(-1, 0, 0, 1, width, 0);
                        break;
                    case 3:
                        ctx.transform(-1, 0, 0, -1, width, height);
                        break;
                    case 4:
                        ctx.transform(1, 0, 0, -1, 0, height);
                        break;
                    case 5:
                        ctx.transform(0, 1, 1, 0, 0, 0);
                        break;
                    case 6:
                        ctx.transform(0, 1, -1, 0, height, 0);
                        break;
                    case 7:
                        ctx.transform(0, -1, -1, 0, height, width);
                        break;
                    case 8:
                        ctx.transform(0, -1, 1, 0, 0, width);
                        break;
                    default:
                        break;
                }
            };
        return ImageUtils;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var ImageCropperComponent = (function () {
        function ImageCropperComponent(elementRef, sanitizer, cd, zone) {
            this.elementRef = elementRef;
            this.sanitizer = sanitizer;
            this.cd = cd;
            this.zone = zone;
            this.marginLeft = '0px';
            this.imageVisible = false;
            this.format = 'png';
            this.outputType = 'both';
            this.maintainAspectRatio = true;
            this.aspectRatio = 1;
            this.resizeToWidth = 0;
            this.roundCropper = false;
            this.onlyScaleDown = false;
            this.imageQuality = 92;
            this.autoCrop = true;
            this.cropper = {
                x1: -100,
                y1: -100,
                x2: 10000,
                y2: 10000
            };
            this.imageCropped = new core.EventEmitter();
            this.imageCroppedBase64 = new core.EventEmitter();
            this.imageCroppedFile = new core.EventEmitter();
            this.imageLoaded = new core.EventEmitter();
            this.loadImageFailed = new core.EventEmitter();
            this.initCropper();
        }
        Object.defineProperty(ImageCropperComponent.prototype, "imageFileChanged", {
            set: /**
             * @param {?} file
             * @return {?}
             */ function (file) {
                this.initCropper();
                if (file) {
                    this.loadImageFile(file);
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ImageCropperComponent.prototype, "imageChangedEvent", {
            set: /**
             * @param {?} event
             * @return {?}
             */ function (event) {
                this.initCropper();
                if (event && event.target && event.target.files && event.target.files.length > 0) {
                    this.loadImageFile(event.target.files[0]);
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ImageCropperComponent.prototype, "imageBase64", {
            set: /**
             * @param {?} imageBase64
             * @return {?}
             */ function (imageBase64) {
                this.initCropper();
                this.loadBase64Image(imageBase64);
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} changes
         * @return {?}
         */
        ImageCropperComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                var _this = this;
                if (changes['cropper']) {
                    setTimeout(function () {
                        _this.setMaxSize();
                        _this.checkCropperPosition(false);
                        _this.doAutoCrop();
                        _this.cd.markForCheck();
                    });
                }
                if (changes['aspectRatio']) {
                    this.resetCropperPosition();
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.initCropper = /**
         * @return {?}
         */
            function () {
                this.imageVisible = false;
                this.originalImage = null;
                this.safeImgDataUrl = 'data:image/png;base64,iVBORw0KGg'
                    + 'oAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAC0lEQVQYV2NgAAIAAAU'
                    + 'AAarVyFEAAAAASUVORK5CYII=';
                this.moveStart = {
                    active: false,
                    type: null,
                    position: null,
                    x1: 0,
                    y1: 0,
                    x2: 0,
                    y2: 0,
                    clientX: 0,
                    clientY: 0
                };
                this.maxSize = {
                    width: 0,
                    height: 0
                };
                this.originalSize = {
                    width: 0,
                    height: 0
                };
                this.cropper.x1 = -100;
                this.cropper.y1 = -100;
                this.cropper.x2 = 10000;
                this.cropper.y2 = 10000;
            };
        /**
         * @param {?} file
         * @return {?}
         */
        ImageCropperComponent.prototype.loadImageFile = /**
         * @param {?} file
         * @return {?}
         */
            function (file) {
                var _this = this;
                /** @type {?} */
                var fileReader = new FileReader();
                fileReader.onload = function (event) {
                    /** @type {?} */
                    var imageType = file.type;
                    if (_this.isValidImageType(imageType)) {
                        try {
                            _this.checkExifRotationAndLoadImage(event.target.result);
                        }
                        catch (e) {
                            _this.loadImageFailed.emit();
                        }
                    }
                    else {
                        _this.loadImageFailed.emit();
                    }
                };
                fileReader.readAsDataURL(file);
            };
        /**
         * @param {?} type
         * @return {?}
         */
        ImageCropperComponent.prototype.isValidImageType = /**
         * @param {?} type
         * @return {?}
         */
            function (type) {
                return type === 'image/jpeg'
                    || type === 'image/jpg'
                    || type === 'image/png'
                    || type === 'image/gif'
                    || type === 'image/tiff';
            };
        /**
         * @param {?} imageBase64
         * @return {?}
         */
        ImageCropperComponent.prototype.checkExifRotationAndLoadImage = /**
         * @param {?} imageBase64
         * @return {?}
         */
            function (imageBase64) {
                var _this = this;
                /** @type {?} */
                var exifRotation = ImageUtils.getOrientation(imageBase64);
                if (exifRotation > 1) {
                    ImageUtils.resetOrientation(imageBase64, exifRotation, function (rotatedBase64) { return _this.loadBase64Image(rotatedBase64); });
                }
                else {
                    this.loadBase64Image(imageBase64);
                }
            };
        /**
         * @param {?} imageBase64
         * @return {?}
         */
        ImageCropperComponent.prototype.loadBase64Image = /**
         * @param {?} imageBase64
         * @return {?}
         */
            function (imageBase64) {
                var _this = this;
                this.safeImgDataUrl = this.sanitizer.bypassSecurityTrustResourceUrl(imageBase64);
                this.originalImage = new Image();
                this.originalImage.onload = function () {
                    _this.originalSize.width = _this.originalImage.width;
                    _this.originalSize.height = _this.originalImage.height;
                    _this.cd.markForCheck();
                };
                this.originalImage.src = imageBase64;
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.imageLoadedInView = /**
         * @return {?}
         */
            function () {
                var _this = this;
                if (this.originalImage != null) {
                    this.imageLoaded.emit();
                    setTimeout(function () {
                        _this.setMaxSize();
                        _this.resetCropperPosition();
                        _this.cd.markForCheck();
                    });
                }
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ImageCropperComponent.prototype.onResize = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.resizeCropperPosition();
                this.setMaxSize();
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.resizeCropperPosition = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var displayedImage = this.elementRef.nativeElement.querySelector('.source-image');
                if (this.maxSize.width !== displayedImage.offsetWidth || this.maxSize.height !== displayedImage.offsetHeight) {
                    this.cropper.x1 = this.cropper.x1 * displayedImage.offsetWidth / this.maxSize.width;
                    this.cropper.x2 = this.cropper.x2 * displayedImage.offsetWidth / this.maxSize.width;
                    this.cropper.y1 = this.cropper.y1 * displayedImage.offsetHeight / this.maxSize.height;
                    this.cropper.y2 = this.cropper.y2 * displayedImage.offsetHeight / this.maxSize.height;
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.resetCropperPosition = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var displayedImage = this.elementRef.nativeElement.querySelector('.source-image');
                if (displayedImage.offsetWidth / this.aspectRatio < displayedImage.offsetHeight) {
                    this.cropper.x1 = 0;
                    this.cropper.x2 = displayedImage.offsetWidth;
                    /** @type {?} */
                    var cropperHeight = displayedImage.offsetWidth / this.aspectRatio;
                    this.cropper.y1 = (displayedImage.offsetHeight - cropperHeight) / 2;
                    this.cropper.y2 = this.cropper.y1 + cropperHeight;
                }
                else {
                    this.cropper.y1 = 0;
                    this.cropper.y2 = displayedImage.offsetHeight;
                    /** @type {?} */
                    var cropperWidth = displayedImage.offsetHeight * this.aspectRatio;
                    this.cropper.x1 = (displayedImage.offsetWidth - cropperWidth) / 2;
                    this.cropper.x2 = this.cropper.x1 + cropperWidth;
                }
                this.doAutoCrop();
                this.imageVisible = true;
            };
        /**
         * @param {?} event
         * @param {?} moveType
         * @param {?=} position
         * @return {?}
         */
        ImageCropperComponent.prototype.startMove = /**
         * @param {?} event
         * @param {?} moveType
         * @param {?=} position
         * @return {?}
         */
            function (event, moveType, position) {
                if (position === void 0) {
                    position = null;
                }
                this.moveStart = Object.assign({
                    active: true,
                    type: moveType,
                    position: position,
                    clientX: this.getClientX(event),
                    clientY: this.getClientY(event)
                }, this.cropper);
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ImageCropperComponent.prototype.moveImg = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (this.moveStart.active) {
                    event.stopPropagation();
                    event.preventDefault();
                    this.setMaxSize();
                    if (this.moveStart.type === 'move') {
                        this.move(event);
                        this.checkCropperPosition(true);
                    }
                    else if (this.moveStart.type === 'resize') {
                        this.resize(event);
                        this.checkCropperPosition(false);
                    }
                    this.cd.detectChanges();
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.setMaxSize = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var el = this.elementRef.nativeElement.querySelector('.source-image');
                this.maxSize.width = el.offsetWidth;
                this.maxSize.height = el.offsetHeight;
                this.marginLeft = this.sanitizer.bypassSecurityTrustStyle('calc(50% - ' + this.maxSize.width / 2 + 'px)');
            };
        /**
         * @param {?=} maintainSize
         * @return {?}
         */
        ImageCropperComponent.prototype.checkCropperPosition = /**
         * @param {?=} maintainSize
         * @return {?}
         */
            function (maintainSize) {
                if (maintainSize === void 0) {
                    maintainSize = false;
                }
                if (this.cropper.x1 < 0) {
                    this.cropper.x2 -= maintainSize ? this.cropper.x1 : 0;
                    this.cropper.x1 = 0;
                }
                if (this.cropper.y1 < 0) {
                    this.cropper.y2 -= maintainSize ? this.cropper.y1 : 0;
                    this.cropper.y1 = 0;
                }
                if (this.cropper.x2 > this.maxSize.width) {
                    this.cropper.x1 -= maintainSize ? (this.cropper.x2 - this.maxSize.width) : 0;
                    this.cropper.x2 = this.maxSize.width;
                }
                if (this.cropper.y2 > this.maxSize.height) {
                    this.cropper.y1 -= maintainSize ? (this.cropper.y2 - this.maxSize.height) : 0;
                    this.cropper.y2 = this.maxSize.height;
                }
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ImageCropperComponent.prototype.moveStop = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (this.moveStart.active) {
                    this.moveStart.active = false;
                    this.doAutoCrop();
                }
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ImageCropperComponent.prototype.move = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                /** @type {?} */
                var diffX = this.getClientX(event) - this.moveStart.clientX;
                /** @type {?} */
                var diffY = this.getClientY(event) - this.moveStart.clientY;
                this.cropper.x1 = this.moveStart.x1 + diffX;
                this.cropper.y1 = this.moveStart.y1 + diffY;
                this.cropper.x2 = this.moveStart.x2 + diffX;
                this.cropper.y2 = this.moveStart.y2 + diffY;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ImageCropperComponent.prototype.resize = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                /** @type {?} */
                var diffX = this.getClientX(event) - this.moveStart.clientX;
                /** @type {?} */
                var diffY = this.getClientY(event) - this.moveStart.clientY;
                switch (this.moveStart.position) {
                    case 'left':
                        this.cropper.x1 = Math.min(this.moveStart.x1 + diffX, this.cropper.x2 - 20);
                        break;
                    case 'topleft':
                        this.cropper.x1 = Math.min(this.moveStart.x1 + diffX, this.cropper.x2 - 20);
                        this.cropper.y1 = Math.min(this.moveStart.y1 + diffY, this.cropper.y2 - 20);
                        break;
                    case 'top':
                        this.cropper.y1 = Math.min(this.moveStart.y1 + diffY, this.cropper.y2 - 20);
                        break;
                    case 'topright':
                        this.cropper.x2 = Math.max(this.moveStart.x2 + diffX, this.cropper.x1 + 20);
                        this.cropper.y1 = Math.min(this.moveStart.y1 + diffY, this.cropper.y2 - 20);
                        break;
                    case 'right':
                        this.cropper.x2 = Math.max(this.moveStart.x2 + diffX, this.cropper.x1 + 20);
                        break;
                    case 'bottomright':
                        this.cropper.x2 = Math.max(this.moveStart.x2 + diffX, this.cropper.x1 + 20);
                        this.cropper.y2 = Math.max(this.moveStart.y2 + diffY, this.cropper.y1 + 20);
                        break;
                    case 'bottom':
                        this.cropper.y2 = Math.max(this.moveStart.y2 + diffY, this.cropper.y1 + 20);
                        break;
                    case 'bottomleft':
                        this.cropper.x1 = Math.min(this.moveStart.x1 + diffX, this.cropper.x2 - 20);
                        this.cropper.y2 = Math.max(this.moveStart.y2 + diffY, this.cropper.y1 + 20);
                        break;
                }
                if (this.maintainAspectRatio) {
                    this.checkAspectRatio();
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.checkAspectRatio = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var overflowX = 0;
                /** @type {?} */
                var overflowY = 0;
                switch (this.moveStart.position) {
                    case 'top':
                        this.cropper.x2 = this.cropper.x1 + (this.cropper.y2 - this.cropper.y1) * this.aspectRatio;
                        overflowX = Math.max(this.cropper.x2 - this.maxSize.width, 0);
                        overflowY = Math.max(0 - this.cropper.y1, 0);
                        if (overflowX > 0 || overflowY > 0) {
                            this.cropper.x2 -= (overflowY * this.aspectRatio) > overflowX ? (overflowY * this.aspectRatio) : overflowX;
                            this.cropper.y1 += (overflowY * this.aspectRatio) > overflowX ? overflowY : overflowX / this.aspectRatio;
                        }
                        break;
                    case 'bottom':
                        this.cropper.x2 = this.cropper.x1 + (this.cropper.y2 - this.cropper.y1) * this.aspectRatio;
                        overflowX = Math.max(this.cropper.x2 - this.maxSize.width, 0);
                        overflowY = Math.max(this.cropper.y2 - this.maxSize.height, 0);
                        if (overflowX > 0 || overflowY > 0) {
                            this.cropper.x2 -= (overflowY * this.aspectRatio) > overflowX ? (overflowY * this.aspectRatio) : overflowX;
                            this.cropper.y2 -= (overflowY * this.aspectRatio) > overflowX ? overflowY : (overflowX / this.aspectRatio);
                        }
                        break;
                    case 'topleft':
                        this.cropper.y1 = this.cropper.y2 - (this.cropper.x2 - this.cropper.x1) / this.aspectRatio;
                        overflowX = Math.max(0 - this.cropper.x1, 0);
                        overflowY = Math.max(0 - this.cropper.y1, 0);
                        if (overflowX > 0 || overflowY > 0) {
                            this.cropper.x1 += (overflowY * this.aspectRatio) > overflowX ? (overflowY * this.aspectRatio) : overflowX;
                            this.cropper.y1 += (overflowY * this.aspectRatio) > overflowX ? overflowY : overflowX / this.aspectRatio;
                        }
                        break;
                    case 'topright':
                        this.cropper.y1 = this.cropper.y2 - (this.cropper.x2 - this.cropper.x1) / this.aspectRatio;
                        overflowX = Math.max(this.cropper.x2 - this.maxSize.width, 0);
                        overflowY = Math.max(0 - this.cropper.y1, 0);
                        if (overflowX > 0 || overflowY > 0) {
                            this.cropper.x2 -= (overflowY * this.aspectRatio) > overflowX ? (overflowY * this.aspectRatio) : overflowX;
                            this.cropper.y1 += (overflowY * this.aspectRatio) > overflowX ? overflowY : overflowX / this.aspectRatio;
                        }
                        break;
                    case 'right':
                    case 'bottomright':
                        this.cropper.y2 = this.cropper.y1 + (this.cropper.x2 - this.cropper.x1) / this.aspectRatio;
                        overflowX = Math.max(this.cropper.x2 - this.maxSize.width, 0);
                        overflowY = Math.max(this.cropper.y2 - this.maxSize.height, 0);
                        if (overflowX > 0 || overflowY > 0) {
                            this.cropper.x2 -= (overflowY * this.aspectRatio) > overflowX ? (overflowY * this.aspectRatio) : overflowX;
                            this.cropper.y2 -= (overflowY * this.aspectRatio) > overflowX ? overflowY : overflowX / this.aspectRatio;
                        }
                        break;
                    case 'left':
                    case 'bottomleft':
                        this.cropper.y2 = this.cropper.y1 + (this.cropper.x2 - this.cropper.x1) / this.aspectRatio;
                        overflowX = Math.max(0 - this.cropper.x1, 0);
                        overflowY = Math.max(this.cropper.y2 - this.maxSize.height, 0);
                        if (overflowX > 0 || overflowY > 0) {
                            this.cropper.x1 += (overflowY * this.aspectRatio) > overflowX ? (overflowY * this.aspectRatio) : overflowX;
                            this.cropper.y2 -= (overflowY * this.aspectRatio) > overflowX ? overflowY : overflowX / this.aspectRatio;
                        }
                        break;
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.doAutoCrop = /**
         * @return {?}
         */
            function () {
                if (this.autoCrop) {
                    this.crop();
                }
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.crop = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var displayedImage = this.elementRef.nativeElement.querySelector('.source-image');
                if (displayedImage && this.originalImage != null) {
                    /** @type {?} */
                    var ratio = this.originalSize.width / displayedImage.offsetWidth;
                    /** @type {?} */
                    var left = Math.round(this.cropper.x1 * ratio);
                    /** @type {?} */
                    var top_1 = Math.round(this.cropper.y1 * ratio);
                    /** @type {?} */
                    var width = Math.min(Math.round((this.cropper.x2 - this.cropper.x1) * ratio), this.originalSize.width - left);
                    /** @type {?} */
                    var height = Math.min(Math.round((this.cropper.y2 - this.cropper.y1) * ratio), this.originalSize.height - top_1);
                    /** @type {?} */
                    var resizeRatio = this.getResizeRatio(width);
                    /** @type {?} */
                    var resizedWidth = Math.floor(width * resizeRatio);
                    /** @type {?} */
                    var resizedHeight = Math.floor(height * resizeRatio);
                    /** @type {?} */
                    var cropCanvas = (document.createElement('canvas'));
                    cropCanvas.width = resizedWidth;
                    cropCanvas.height = resizedHeight;
                    /** @type {?} */
                    var ctx = cropCanvas.getContext('2d');
                    if (ctx) {
                        ctx.drawImage(this.originalImage, left, top_1, width, height, 0, 0, resizedWidth, resizedHeight);
                        this.cropToOutputType(cropCanvas, resizedWidth, resizedHeight);
                    }
                }
            };
        /**
         * @param {?} cropCanvas
         * @param {?} resizedWidth
         * @param {?} resizedHeight
         * @return {?}
         */
        ImageCropperComponent.prototype.cropToOutputType = /**
         * @param {?} cropCanvas
         * @param {?} resizedWidth
         * @param {?} resizedHeight
         * @return {?}
         */
            function (cropCanvas, resizedWidth, resizedHeight) {
                var _this = this;
                /** @type {?} */
                var output = {
                    width: resizedWidth,
                    height: resizedHeight,
                    cropperPosition: Object.assign({}, this.cropper)
                };
                switch (this.outputType) {
                    case 'file':
                        return this.cropToFile(cropCanvas)
                            .then(function (result) {
                            output.file = result;
                            _this.imageCropped.emit(output);
                            return output;
                        });
                    case 'both':
                        output.base64 = this.cropToBase64(cropCanvas);
                        return this.cropToFile(cropCanvas)
                            .then(function (result) {
                            output.file = result;
                            _this.imageCropped.emit(output);
                            return output;
                        });
                    default:
                        output.base64 = this.cropToBase64(cropCanvas);
                        this.imageCropped.emit(output);
                        return output;
                }
            };
        /**
         * @param {?} cropCanvas
         * @return {?}
         */
        ImageCropperComponent.prototype.cropToBase64 = /**
         * @param {?} cropCanvas
         * @return {?}
         */
            function (cropCanvas) {
                /** @type {?} */
                var imageBase64 = cropCanvas.toDataURL('image/' + this.format, this.getQuality());
                this.imageCroppedBase64.emit(imageBase64);
                return imageBase64;
            };
        /**
         * @param {?} cropCanvas
         * @return {?}
         */
        ImageCropperComponent.prototype.cropToFile = /**
         * @param {?} cropCanvas
         * @return {?}
         */
            function (cropCanvas) {
                var _this = this;
                return this.getCanvasBlob(cropCanvas)
                    .then(function (result) {
                    if (result) {
                        _this.imageCroppedFile.emit(result);
                    }
                    return result;
                });
            };
        /**
         * @param {?} cropCanvas
         * @return {?}
         */
        ImageCropperComponent.prototype.getCanvasBlob = /**
         * @param {?} cropCanvas
         * @return {?}
         */
            function (cropCanvas) {
                var _this = this;
                return new Promise(function (resolve) {
                    cropCanvas.toBlob(function (result) { return _this.zone.run(function () { return resolve(result); }); }, 'image/' + _this.format, _this.getQuality());
                });
            };
        /**
         * @return {?}
         */
        ImageCropperComponent.prototype.getQuality = /**
         * @return {?}
         */
            function () {
                return Math.min(1, Math.max(0, this.imageQuality / 100));
            };
        /**
         * @param {?} width
         * @return {?}
         */
        ImageCropperComponent.prototype.getResizeRatio = /**
         * @param {?} width
         * @return {?}
         */
            function (width) {
                return this.resizeToWidth > 0 && (!this.onlyScaleDown || width > this.resizeToWidth)
                    ? this.resizeToWidth / width
                    : 1;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ImageCropperComponent.prototype.getClientX = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                return event.clientX != null ? event.clientX : event.touches[0].clientX;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        ImageCropperComponent.prototype.getClientY = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                return event.clientY != null ? event.clientY : event.touches[0].clientY;
            };
        ImageCropperComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'image-cropper',
                        template: "<div>\n    <img\n        [src]=\"safeImgDataUrl\"\n        [style.visibility]=\"imageVisible ? 'visible' : 'hidden'\"\n        (load)=\"imageLoadedInView()\"\n        class=\"source-image\"\n    />\n    <div class=\"cropper\"\n         [class.rounded]=\"roundCropper\"\n         [style.top.px]=\"cropper.y1\"\n         [style.left.px]=\"cropper.x1\"\n         [style.width.px]=\"cropper.x2 - cropper.x1\"\n         [style.height.px]=\"cropper.y2 - cropper.y1\"\n         [style.margin-left]=\"marginLeft\"\n         [style.visibility]=\"imageVisible ? 'visible' : 'hidden'\"\n    >\n        <div\n            (mousedown)=\"startMove($event, 'move')\"\n            (touchstart)=\"startMove($event, 'move')\"\n            class=\"move\"\n        >&nbsp;</div>\n        <span\n            class=\"resize topleft\"\n            (mousedown)=\"startMove($event, 'resize', 'topleft')\"\n            (touchstart)=\"startMove($event, 'resize', 'topleft')\"\n        ><span class=\"square\"></span></span>\n        <span\n            class=\"resize top\"\n        ><span class=\"square\"></span></span>\n        <span\n            class=\"resize topright\"\n            (mousedown)=\"startMove($event, 'resize', 'topright')\"\n            (touchstart)=\"startMove($event, 'resize', 'topright')\"\n        ><span class=\"square\"></span></span>\n        <span\n            class=\"resize right\"\n        ><span class=\"square\"></span></span>\n        <span\n            class=\"resize bottomright\"\n            (mousedown)=\"startMove($event, 'resize', 'bottomright')\"\n            (touchstart)=\"startMove($event, 'resize', 'bottomright')\"\n        ><span class=\"square\"></span></span>\n        <span\n            class=\"resize bottom\"\n        ><span class=\"square\"></span></span>\n        <span\n            class=\"resize bottomleft\"\n            (mousedown)=\"startMove($event, 'resize', 'bottomleft')\"\n            (touchstart)=\"startMove($event, 'resize', 'bottomleft')\"\n        ><span class=\"square\"></span></span>\n        <span\n            class=\"resize left\"\n        ><span class=\"square\"></span></span>\n        <span\n            class=\"resize-bar top\"\n            (mousedown)=\"startMove($event, 'resize', 'top')\"\n            (touchstart)=\"startMove($event, 'resize', 'top')\"\n        ></span>\n        <span\n            class=\"resize-bar right\"\n            (mousedown)=\"startMove($event, 'resize', 'right')\"\n            (touchstart)=\"startMove($event, 'resize', 'right')\"\n        ></span>\n        <span\n            class=\"resize-bar bottom\"\n            (mousedown)=\"startMove($event, 'resize', 'bottom')\"\n            (touchstart)=\"startMove($event, 'resize', 'bottom')\"\n        ></span>\n        <span\n            class=\"resize-bar left\"\n            (mousedown)=\"startMove($event, 'resize', 'left')\"\n            (touchstart)=\"startMove($event, 'resize', 'left')\"\n        ></span>\n    </div>\n</div>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        styles: [":host{display:flex;position:relative;width:100%;max-width:100%;max-height:100%;overflow:hidden;padding:5px;text-align:center;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}:host>div{position:relative;width:100%}:host>div .source-image{max-width:100%;max-height:100%}:host .cropper{position:absolute;display:flex;color:#53535c!important;background:0 0!important;touch-action:none;outline:rgba(255,255,255,.3) solid 1000px}:host .cropper:after{position:absolute;content:'';top:0;bottom:0;left:0;right:0;pointer-events:none;border:1px dashed;opacity:.75;color:inherit;z-index:1}:host .cropper .move{width:100%;cursor:move;border:1px solid rgba(255,255,255,.5)}:host .cropper .resize{position:absolute;display:inline-block;line-height:6px;padding:8px;opacity:.85;z-index:1}:host .cropper .resize .square{display:inline-block;background:#53535c!important;width:6px;height:6px;border:1px solid rgba(255,255,255,.5)}:host .cropper .resize.topleft{top:-12px;left:-12px;cursor:nw-resize}:host .cropper .resize.top{top:-12px;left:calc(50% - 12px);cursor:n-resize}:host .cropper .resize.topright{top:-12px;right:-12px;cursor:ne-resize}:host .cropper .resize.right{top:calc(50% - 12px);right:-12px;cursor:e-resize}:host .cropper .resize.bottomright{bottom:-12px;right:-12px;cursor:se-resize}:host .cropper .resize.bottom{bottom:-12px;left:calc(50% - 12px);cursor:s-resize}:host .cropper .resize.bottomleft{bottom:-12px;left:-12px;cursor:sw-resize}:host .cropper .resize.left{top:calc(50% - 12px);left:-12px;cursor:w-resize}:host .cropper .resize-bar{position:absolute;z-index:1}:host .cropper .resize-bar.top{top:-11px;left:11px;width:calc(100% - 22px);height:22px;cursor:n-resize}:host .cropper .resize-bar.right{top:11px;right:-11px;height:calc(100% - 22px);width:22px;cursor:e-resize}:host .cropper .resize-bar.bottom{bottom:-11px;left:11px;width:calc(100% - 22px);height:22px;cursor:s-resize}:host .cropper .resize-bar.left{top:11px;left:-11px;height:calc(100% - 22px);width:22px;cursor:w-resize}:host .cropper.rounded{outline-color:transparent}:host .cropper.rounded:after{box-shadow:0 0 0 100vw rgba(255,255,255,.3);border-radius:100%}:host .cropper.rounded .move{border-radius:100%}"]
                    }] }
        ];
        /** @nocollapse */
        ImageCropperComponent.ctorParameters = function () {
            return [
                { type: core.ElementRef },
                { type: platformBrowser.DomSanitizer },
                { type: core.ChangeDetectorRef },
                { type: core.NgZone }
            ];
        };
        ImageCropperComponent.propDecorators = {
            imageFileChanged: [{ type: core.Input }],
            imageChangedEvent: [{ type: core.Input }],
            imageBase64: [{ type: core.Input }],
            format: [{ type: core.Input }],
            outputType: [{ type: core.Input }],
            maintainAspectRatio: [{ type: core.Input }],
            aspectRatio: [{ type: core.Input }],
            resizeToWidth: [{ type: core.Input }],
            roundCropper: [{ type: core.Input }],
            onlyScaleDown: [{ type: core.Input }],
            imageQuality: [{ type: core.Input }],
            autoCrop: [{ type: core.Input }],
            cropper: [{ type: core.Input }],
            imageCropped: [{ type: core.Output }],
            imageCroppedBase64: [{ type: core.Output }],
            imageCroppedFile: [{ type: core.Output }],
            imageLoaded: [{ type: core.Output }],
            loadImageFailed: [{ type: core.Output }],
            onResize: [{ type: core.HostListener, args: ['window:resize', ['$event'],] }],
            moveImg: [{ type: core.HostListener, args: ['document:mousemove', ['$event'],] }, { type: core.HostListener, args: ['document:touchmove', ['$event'],] }],
            moveStop: [{ type: core.HostListener, args: ['document:mouseup', ['$event'],] }, { type: core.HostListener, args: ['document:touchend', ['$event'],] }]
        };
        return ImageCropperComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var ImageCropperModule = (function () {
        function ImageCropperModule() {
        }
        ImageCropperModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [
                            ImageCropperComponent
                        ],
                        exports: [
                            ImageCropperComponent
                        ]
                    },] }
        ];
        return ImageCropperModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.ImageCropperModule = ImageCropperModule;
    exports.ImageCropperComponent = ImageCropperComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
